### PR TITLE
refactor: let parameters accept argc and argv

### PR DESCRIPTION
**Pull request by d3adb5**
Let parameters accept an argument count and array of argument values instead of a single argument. The parameter struct has been changed to allow a range of argument counts instead of the previous, takes & is_optional implementation.

The rationale is that after this change, we can have this kind of thing in the rc file:

```
ignore_classes fcitx steam spotify
```

### CODE REVIEW

**Comment by d3adb5** [comment]
I took the liberty of allowing for lines with over 80 characters and using C99 comments. I might be missing something, but I didn't find any unit tests I could run to test this kind of change. It's a hefty PR for a single commit, I'd appreciate some feedback, @kolbusa.

**Comment by kolbusa** [comment]
Thanks. Let me look over the weekend. I would prefer keeping stuff to C89 as using anything newer used to bite me previously. Formatting can be 'fixed' via clang-format. The `.clang-format` in the top level directory is up to date. You will probably get some spurious additional edits due to some older formatting errors, and that's OK. I will add clang-format check into the CI when I have time.

**Comment by kolbusa** [comment]
I am testing this branch on my machine and see difference in icon placement with and without the change. I did not have time to look deeper, but can you please double check that for example the default config file parses identically?

**Comment by d3adb5** [comment]
I'll have a look in the coming days. In the meantime, what's different about the icon placement compared to current behavior? And have you noticed any other changes from default behavior?

**Comment by kolbusa** [comment]
It looks like the `icon_gravity` is parsed incorrectly. I have `SE`, which is `0x5`, but the result is `0xa` which is `NW`. 

Config:
```
decorations none
geometry 6x1
icon_size 18
max_geometry 6x1
parent_bg true
transparent false
icon_gravity SE
no_shrink true
```

Regarding formatting: I would like to keep the code C89-compatible and aligned with the `.clang-format`. I think I should add a CI check for that...

**Comment by d3adb5** [comment]
> It looks like the `icon_gravity` is parsed incorrectly. I have `SE`, which is `0x5`, but the result is `0xa` which is `NW`.

It was indeed. `parse_gravity()` wasn't actually changing the setting passed through `references` (previously `tgt`). I fixed that in 90b5f42.

> I would like to keep the code C89-compatible and aligned with the `.clang-format`. I think I should add a CI check for that...

I understand. I know compound literals like there were before aren't part of C89, but they're available as an extension. I just wonder whether the `{ .member1 = value1, .member2 = value2 }` syntax will be valid. I'm going to run clang-format on the files and see if it still compiles.

**Comment by d3adb5** [comment]
@kolbusa I ran `clang-format`, but since it's configured for C++03 (and I couldn't find how to make it work for C89), it didn't eliminate the C99 style comments. It did, however, change **many** lines outside the ones this PR changes. Are you able to configure it for C89 somehow? For the time being I'm going to rever to C89-style comments.

Let me know if I can help you with the CI check, too!

**Comment by kolbusa** [comment]
I see no issues in my (limited) testing so far. I should probably do a reformatting pass after this PR is merged. The check should be just a single line in the CI yml, but thanks for offering to help!

File: src/settings.c [hunk 1]
```diff
@@ -97,254 +97,301 @@ void init_default_settings()
...
-int parse_scrollbars_highlight_color(char *str, char ***tgt, int silent)
+int parse_scrollbars_highlight_color(int argc, const char **argv, void **references, int silent)
 {
-    if (!strcasecmp(str, "disable"))
-        **tgt = NULL;
-    else {
-        **tgt = strdup(str);
-        if (**tgt == NULL) DIE_OOM(("Could not copy value from parameter\n"));
-    }
+    char **highlight_color = (char **) references[0];
```
Comment (kolbusa) [inline src/settings.c:102]: (Random spot, but applies to most of the changes in this PR): I think the new parsing functions much check that `argc == 1` unless they accept multiple arguments.

**Review by kolbusa — CHANGES_REQUESTED** [review]
I left a comment re `argc`. Otherwise LGTM.

File: src/settings.c (hunk 1 above)
Comment (d3adb5) [reply src/settings.c:102]: I thought of validating the number of arguments in the parsing functions as well, especially since the responsibility of parameter validation is more often falls on the function than its callers, but [I embedded the allowed range of argument counts in the `Param` struct.](https://github.com/kolbusa/stalonetray/pull/27/files#diff-b16b9fd7ad76f07b54dd144e7c9917d017b82ab74ee7cd0cba9f7c750e2c509aR508-R509) That range is used for validation [before passing them to the parsing function,](https://github.com/kolbusa/stalonetray/pull/27/files#diff-b16b9fd7ad76f07b54dd144e7c9917d017b82ab74ee7cd0cba9f7c750e2c509aR1546-R1551) so I found it acceptable to delegate that to whatever procedure is reading `struct Param` type values.

Doing otherwise might simplify `void parse_rc()`, but in my opinion will lead to plenty of repetitive checks in each `parse_*` function. Additionally, we would either have to get rid of `int min_argc` and `int max_argc` in the structured type, relying on hardcoded values in each parsing function, or change the parser signature again to give it a `struct Param *` parameter and repeat the check in each parsing function:

```c
int parse_example(int argc, char **argv, struct Param *param, int silent)
{
  /* variable definitions */
  
  if (argc > param->max_argc || argc < param->min_argc) {
    PARSING_ERROR(/* ... */);
    return FAILURE;
  }

  /* ... */
}
```

What do you think is the best way forward? I agree that lack of validation in the parsing functions seems off, but I thought it was convenient to have as much information as possible embedded into the parameter type structure.

To sum up the options I see, none of which I find optimal:

1. Leave as-is, delegating responsibility to `parse_rc()` and `parse_cmdline()`.
2. Remove validation from `parse_rc()` and `parse_cmdline()`, remove argument count constraints from `struct Param`, hardcode valid range when validating `argc` in each parsing function.
3. Remove validation from `parse_rc()` and `parse_cmdline()` and move it into each parsing function, which will now receive a `struct Param *` argument.

**Comment by d3adb5** [comment]
I added a commit that reverts the C99 style comments to C89 compatible ones (8ca0a54). Waiting to hear your opinions on the argument count validation, @kolbusa.

**Review by kolbusa — APPROVED** [review]
